OpenRA.Mods.Common/Lint/CheckTranslationReference.cs

```diff
@@ -226,16 +500,47 @@ void ILintPass.Run(Action<string> emitError, Action<string> emitWarning, ModData
 				if (!key.Any(char.IsLetter))
 					continue;
 
-				if (!translation.HasMessage(key))
-					emitWarning($"`{key}` defined by `{node.Key}` is not present in `{language}` translation.");
+				if (!translation.HasMessage(key) && isFirstBound)
+					emitWarning($"`{key}` defined by `{node.Key}` in field `{childType}` in {node.Location} is not present in `{language}` translation.");
 
 				referencedKeys.Add(key);
+
+				// HACK: Tooltips don't display on the widget directly, don't validate their sizes.
+				if (childType == "TooltipText" || childType == "TooltipDesc")
+					continue;
+
+				// HACK: Hardcode how each widget determines available fonts.
+				var fontName = nodeType switch
+				{
+					"Button" or "DropDownButton" or "Checkbox" or "MenuButton" or "WorldButton" => node.Value.NodeWithKeyOrDefault("Font")?.Value.Value ?? ChromeMetrics.Get<string>("ButtonFont"),
```

Comment:
Would it make sense to add a `FontReference` attribute that we can reflect over instead?

Reply:
Yes, that'd be the scalable approach to solving this. I don't think I ran into anything that uses multiple fonts/texts but yes we could encode that too.

Compared to the relationships, this attribute would be easier to get set up.

Reply:
But not that easy, since we need the default values from the class and we're not actually instantiating the widgets. :/

Reply:
Looking up the translation values everywhere in the classes is a bit messy. Having an underlying layer take care of it would be cleaner. 

The translation is now showing up everywhere in each constructor mixing with logic.

Pseudo code
```
class SomeWidget
{
  struct LocaleResource
  {
   // Would lookup "Scope.SomeWidget.buttonCaption1" from a resource file.
   [LocaleString id="button1Caption"]
   string buttonCaption;
  }

  [LocaleResource id="Scope.SomeWidget"]
  static LocaleResource locale;

  static LocaleResource GetLocale(Mod mod, bool force = false)
  {
    var mod = mod || GetDefaultMod(mod);
     // init once
     return locale;
  }

  void SomeMethod()
  {
     var buttonCaption = GetLocale(GetModFromSomeWhere()).buttonCaption;
  }
}

```

This would then also allow you to load resources without instantiating the class. I.e. `CurentNamespace.Locale.SomeWidgetLocaleResource.cs`.

You could even pull the struct outside the class, put all in a Locale namespace. Which your linter could scan.

Also avoid loading translation with each instantation of i.e. widgets. Loading of widgets is already quite dynamic and slow.

Reply:
I think you've effectively described the existing translation setup? We're already able to pull the translation text successfully without having to instantiate widgets. It's other references like the fonts and widget IDs that don't have such a mechanism and can't easily add one since their data might be set via code. Whereas translation config is defined in YAML only.

I think the outcome here is that the PR as it stands is pretty much what we're going to get for a "quick win". Replacing the hacks with a proper solution would be a non-trivial undertaking. So I'd say review this on the basis of whether you think these value justifies the amount of hacks required.

Reply:
I don't think this is mergable, for next release we need to make sure translations have proper third party support, and this PR will only delay us from reaching that milestone. If noone's gonna be unhardcoding this lint test then I think a second PR should be split from this with the yaml changes

Reply:
Having the lint pass detect some size issues for you still seems to be a net win to me compared to not detecting anything at all. But split out #21309 so we can at least get that merged.